File: mandalay_bay/rewards.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from mandalay_bay.chips import ChipWallet, TransactionKind

if TYPE_CHECKING:
    from mandalay_bay.session import PlayerSession

SAVE_VERSION_WITH_REWARDS = 3

COMP_CATALOG: dict[str, dict[str, str]] = {
    "welcome_drink": {
        "title": "Welcome Cocktail",
        "body": "Complimentary drink at Betty's Bar — enrollment perk.",
    },
    "slot_freeplay": {
        "title": "$10 Slot Free-Play",
        "body": "Pearl tier voucher — flavor credit for the slot aisle.",
    },
    "buffet_comp": {
        "title": "Buffet Comp",
        "body": "Gold tier — the line moves faster when you're comped.",
    },
    "room_night": {
        "title": "Standard Room Night",
        "body": "Platinum tier — one night on the house (narrative comp).",
    },
    "suite_upgrade": {
        "title": "Suite Upgrade",
        "body": "Noir tier — VIP lounge access flag unlocked.",
    },
    "penthouse_fantasy": {
        "title": "Penthouse Fantasy Comp",
        "body": "Chairman tier — the penthouse whispered your name.",
    },
}


@dataclass(frozen=True, slots=True)
class TierDef:
    id: str
    label: str
    min_wagered: int
    comp: str | None = None


TIERS: tuple[TierDef, ...] = (
    TierDef("sapphire", "Sapphire", 0),
    TierDef("pearl", "Pearl", 500, "slot_freeplay"),
    TierDef("gold", "Gold", 2000, "buffet_comp"),
    TierDef("platinum", "Platinum", 5000, "room_night"),
    TierDef("noir", "Noir", 10000, "suite_upgrade"),
    TierDef("chairman", "Chairman", 25000, "penthouse_fantasy"),
)


@dataclass
class RewardsState:
    member_id: str = "MB-000000"
    tier: str = "sapphire"
    lifetime_wagered: int = 0
    unlocked_comps: list[str] = field(default_factory=lambda: ["welcome_drink"])
    redeemed_comps: list[str] = field(default_factory=list)
    notifications: list[dict] = field(default_factory=list)


def default_rewards_state() -> RewardsState:
    return RewardsState(
        notifications=[{
            "id": "welcome_enroll",
            "title": "Welcome to MGM Rewards",
            "body": "Your Sapphire membership is active.",
            "read": False,
        }]
    )


def total_wagered_from_wallet(wallet: ChipWallet) -> int:
    return sum(
        abs(t.amount)
        for t in wallet.transactions
        if t.kind == TransactionKind.WAGER
    )


def tier_for_wagered(amount: int) -> TierDef:
    current = TIERS[0]
    for tier in TIERS:
        if amount >= tier.min_wagered:
            current = tier
    return current


def next_tier(current_tier_id: str) -> TierDef | None:
    for idx, tier in enumerate(TIERS):
        if tier.id == current_tier_id and idx < len(TIERS) - 1:
            return TIERS[idx + 1]
    return None


def ensure_rewards(session: PlayerSession) -> RewardsState:
    if not hasattr(session, "rewards") or session.rewards is None:
        session.rewards = default_rewards_state()
    return session.rewards


def has_unredeemed_comp(rewards: RewardsState, comp_id: str) -> bool:
    return comp_id in rewards.unlocked_comps and comp_id not in rewards.redeemed_comps


def redeem_comp(rewards: RewardsState, comp_id: str) -> bool:
    if not has_unredeemed_comp(rewards, comp_id):
        return False
    rewards.redeemed_comps.append(comp_id)
    meta = COMP_CATALOG.get(comp_id, {})
    rewards.notifications.insert(0, {
        "title": "Comp Redeemed",
        "body": f"{meta.get('title', comp_id)} — enjoy, member.",
        "read": False,
    })
    return True


def progress_to_next_tier(rewards: RewardsState) -> dict:
    nxt = next_tier(rewards.tier)
    if nxt is None:
        return {"pct": 100, "next": None, "remaining": 0}
    current = tier_for_wagered(rewards.lifetime_wagered)
    span = nxt.min_wagered - current.min_wagered
    into = rewards.lifetime_wagered - current.min_wagered
    pct = min(100, round((into / span) * 100)) if span > 0 else 100
    return {"pct": pct, "next": nxt, "remaining": max(0, nxt.min_wagered - rewards.lifetime_wagered)}


def notify_tier_promotion(session_rewards: RewardsState, prev: int, new: int) -> list[str]:
    """Return titles of new tier notifications (CLI flavor)."""
    prev_tier = tier_for_wagered(prev)
    new_tier = tier_for_wagered(new)
    if new_tier.id == prev_tier.id:
        return []
    session_rewards.tier = new_tier.id
    if new_tier.comp and new_tier.comp not in session_rewards.unlocked_comps:
        session_rewards.unlocked_comps.append(new_tier.comp)
    comp_label = COMP_CATALOG.get(new_tier.comp or "", {}).get("title", "exclusive offers")
    from mandalay_bay.rewards_perks import get_tier_experience

    exp = get_tier_experience(new_tier.id)
    title = f"{new_tier.label} Tier Unlocked!"
    body = f"You've reached {new_tier.label} status."
    if new_tier.comp:
        body += f" New comp: {comp_label}."
    body += f" {exp.tagline}"
    session_rewards.notifications.insert(0, {"title": title, "body": body, "read": False})
    return [title]


def sync_rewards_from_wallet(session: PlayerSession) -> list[str]:
    """Sync lifetime wagered and tier comps from wallet ledger. Returns new notification titles."""
    rewards = ensure_rewards(session)
    total = total_wagered_from_wallet(session.wallet)
    prev = rewards.lifetime_wagered
    if total <= prev:
        return []
    rewards.lifetime_wagered = total
    return notify_tier_promotion(rewards, prev, total)


def migrate_session_rewards(session: PlayerSession, data_version: int) -> None:
    ensure_rewards(session)
    total = total_wagered_from_wallet(session.wallet)
    rewards = session.rewards
    assert rewards is not None
    rewards.lifetime_wagered = max(rewards.lifetime_wagered, total)
    rewards.tier = tier_for_wagered(rewards.lifetime_wagered).id
    if data_version < SAVE_VERSION_WITH_REWARDS and not rewards.notifications:
        rewards.notifications.append({
            "id": "welcome_enroll",
            "title": "Welcome to MGM Rewards",
            "body": "Your Sapphire membership is active. Play to earn comps!",
            "read": False,
        })

File: docs/js/rewards.js
import { TransactionKind, secureRandomInt } from "./core.js";
import { getTierExperience } from "./rewards-perks.js";

export const SAVE_VERSION_WITH_REWARDS = 3;

export const TIERS = [
  { id: "sapphire", label: "Sapphire", minWagered: 0, comp: null },
  { id: "pearl", label: "Pearl", minWagered: 500, comp: "slot_freeplay" },
  { id: "gold", label: "Gold", minWagered: 2000, comp: "buffet_comp" },
  { id: "platinum", label: "Platinum", minWagered: 5000, comp: "room_night" },
  { id: "noir", label: "Noir", minWagered: 10000, comp: "suite_upgrade" },
  { id: "chairman", label: "Chairman", minWagered: 25000, comp: "penthouse_fantasy" },
];

export const COMP_CATALOG = {
  welcome_drink: {
    id: "welcome_drink",
    title: "Welcome Cocktail",
    body: "Complimentary drink at Betty's Bar — enrollment perk.",
    redeemFlag: "redeemed_welcome_drink",
  },
  slot_freeplay: {
    id: "slot_freeplay",
    title: "$10 Slot Free-Play",
    body: "Pearl tier voucher — flavor credit for the slot aisle.",
    redeemFlag: "redeemed_slot_freeplay",
  },
  buffet_comp: {
    id: "buffet_comp",
    title: "Buffet Comp",
    body: "Gold tier — the line moves faster when you're comped.",
    redeemFlag: "redeemed_buffet_comp",
  },
  room_night: {
    id: "room_night",
    title: "Standard Room Night",
    body: "Platinum tier — one night on the house (narrative comp).",
    redeemFlag: "redeemed_room_night",
  },
  suite_upgrade: {
    id: "suite_upgrade",
    title: "Suite Upgrade",
    body: "Noir tier — VIP lounge access flag unlocked.",
    redeemFlag: "redeemed_suite_upgrade",
  },
  penthouse_fantasy: {
    id: "penthouse_fantasy",
    title: "Penthouse Fantasy Comp",
    body: "Chairman tier — the penthouse whispered your name.",
    redeemFlag: "redeemed_penthouse_fantasy",
  },
};

function nowIso() {
  return new Date().toISOString();
}

export function defaultRewardsState(overrides = {}) {
  const memberId = overrides.memberId ?? `MB-${secureRandomInt(100000, 999999)}`;
  const base = {
    memberId,
    tier: "sapphire",
    lifetimeWagered: 0,
    unlockedComps: ["welcome_drink"],
    redeemedComps: [],
    notifications: [],
    phoneBook: { threads: {}, easterEggs: [], introSent: [] },
    ...overrides,
  };
  if (!base.notifications.length) {
    base.notifications.push({
      id: "welcome_enroll",
      title: "Welcome to MGM Rewards",
      body: "Your Sapphire membership is active. Play to earn comps — check Betty's Bar for your welcome cocktail!",
      read: false,
      timestamp: nowIso(),
    });
  }
  return base;
}

export function tierForWagered(amount) {
  let current = TIERS[0];
  for (const tier of TIERS) {
    if (amount >= tier.minWagered) current = tier;
  }
  return current;
}

export function nextTier(currentTierId) {
  const idx = TIERS.findIndex((t) => t.id === currentTierId);
  return idx >= 0 && idx < TIERS.length - 1 ? TIERS[idx + 1] : null;
}

export function totalWageredFromWallet(wallet) {
  if (!wallet?.transactions) return 0;
  return wallet.transactions
    .filter((t) => t.kind === TransactionKind.WAGER)
    .reduce((sum, t) => sum + Math.abs(t.amount), 0);
}

/**
 * Track MGM Rewards tier progress and notifications for a PlayerSession.
 */
export class RewardsTracker {
  /**
   * @param {import("./core.js").PlayerSession} session
   * @param {{ onNotify?: (n: object) => void }} [hooks]
   */
  constructor(session, hooks = {}) {
    this.session = session;
    this.onNotify = hooks.onNotify ?? null;
  }

  ensureRewards() {
    const defaults = defaultRewardsState();
    if (!this.session.rewards) {
      this.session.rewards = defaults;
      return this.session.rewards;
    }
    const rewards = this.session.rewards;
    if (!rewards.memberId) rewards.memberId = defaults.memberId;
    if (!rewards.tier) rewards.tier = defaults.tier;
    if (typeof rewards.lifetimeWagered !== "number") rewards.lifetimeWagered = 0;
    if (!Array.isArray(rewards.unlockedComps)) rewards.unlockedComps = [...defaults.unlockedComps];
    if (!Array.isArray(rewards.redeemedComps)) rewards.redeemedComps = [];
    if (!Array.isArray(rewards.notifications)) rewards.notifications = [...defaults.notifications];
    if (!rewards.phoneBook) {
      rewards.phoneBook = { threads: {}, easterEggs: [], introSent: [] };
    }
    return rewards;
  }

  /** @returns {object[]} newly created notifications */
  syncFromWallet() {
    const rewards = this.ensureRewards();
    const total = totalWageredFromWallet(this.session.wallet);
    const prev = rewards.lifetimeWagered;
    if (total <= prev) return [];
    rewards.lifetimeWagered = total;
    return this._applyTierProgress(rewards, prev, total);
  }

  /** @returns {object[]} */
  _applyTierProgress(rewards, prevTotal, newTotal) {
    const created = [];
    const prevTier = tierForWagered(prevTotal);
    const newTier = tierForWagered(newTotal);
    if (newTier.id !== prevTier.id) {
      rewards.tier = newTier.id;
      if (newTier.comp && !rewards.unlockedComps.includes(newTier.comp)) {
        rewards.unlockedComps.push(newTier.comp);
      }
      const compLabel = newTier.comp ? COMP_CATALOG[newTier.comp]?.title : "exclusive offers";
      const exp = getTierExperience(newTier.id);
      const note = {
        id: `tier_${newTier.id}_${Date.now()}`,
        title: `${newTier.label} Tier Unlocked!`,
        body: `You've reached ${newTier.label} status. ${compLabel ? `New comp: ${compLabel}. ` : ""}${exp.tagline}${newTier.id === "gold" || newTier.id === "platinum" ? " Check Connect on your phone for new contacts!" : ""}`,
        read: false,
        timestamp: nowIso(),
      };
      rewards.notifications.unshift(note);
      created.push(note);
      this.onNotify?.(note);
    }
    return created;
  }

  unreadCount() {
    const rewards = this.ensureRewards();
    return rewards.notifications.filter((n) => !n.read).length;
  }

  markRead(notificationId) {
    const rewards = this.ensureRewards();
    const note = rewards.notifications.find((n) => n.id === notificationId);
    if (note) note.read = true;
  }

  markAllRead() {
    for (const n of this.ensureRewards().notifications) n.read = true;
  }

  redeemComp(compId) {
    const rewards = this.ensureRewards();
    if (!rewards.unlockedComps.includes(compId)) return false;
    if (rewards.redeemedComps.includes(compId)) return false;
    rewards.redeemedComps.push(compId);
    const meta = COMP_CATALOG[compId];
    if (meta?.redeemFlag && this.session.rpg) {
      this.session.rpg.flags = this.session.rpg.flags ?? {};
      this.session.rpg.flags[meta.redeemFlag] = true;
      if (compId === "welcome_drink") {
        delete this.session.rpg.flags.has_welcome_drink_comp;
      }
    }
    const note = {
      id: `redeem_${compId}_${Date.now()}`,
      title: "Comp Redeemed",
      body: `${meta?.title ?? compId} — enjoy, member.`,
      read: false,
      timestamp: nowIso(),
    };
    rewards.notifications.unshift(note);
    this.onNotify?.(note);
    return true;
  }

  pushNotification(title, body, id = null) {
    const rewards = this.ensureRewards();
    const note = {
      id: id ?? `note_${Date.now()}`,
      title,
      body,
      read: false,
      timestamp: nowIso(),
    };
    rewards.notifications.unshift(note);
    this.onNotify?.(note);
    return note;
  }

  progressToNextTier() {
    const rewards = this.ensureRewards();
    const next = nextTier(rewards.tier);
    if (!next) return { pct: 100, next: null, remaining: 0 };
    const currentTier = tierForWagered(rewards.lifetimeWagered);
    const span = next.minWagered - currentTier.minWagered;
    const into = rewards.lifetimeWagered - currentTier.minWagered;
    const pct = span > 0 ? Math.min(100, Math.round((into / span) * 100)) : 100;
    return { pct, next, remaining: Math.max(0, next.minWagered - rewards.lifetimeWagered) };
  }
}

export function migrateSessionRewards(session, dataVersion) {
  if (!session.rewards) {
    session.rewards = defaultRewardsState();
  }
  const total = totalWageredFromWallet(session.wallet);
  session.rewards.lifetimeWagered = Math.max(session.rewards.lifetimeWagered, total);
  session.rewards.tier = tierForWagered(session.rewards.lifetimeWagered).id;
  if (dataVersion < SAVE_VERSION_WITH_REWARDS && !session.rewards.notifications.length) {
    session.rewards.notifications.push({
      id: "welcome_enroll",
      title: "Welcome to MGM Rewards",
      body: "Your Sapphire membership is active. Play to earn comps!",
      read: false,
      timestamp: nowIso(),
    });
  }
  syncRewardsFlags(session);
}

export function syncRewardsFlags(session) {
  if (!session.rpg) return;
  session.rpg.flags = session.rpg.flags ?? {};
  const rewards = session.rewards;
  if (!rewards) return;
  if (rewards.unlockedComps.includes("welcome_drink") && !rewards.redeemedComps.includes("welcome_drink")) {
    session.rpg.flags.has_welcome_drink_comp = true;
  }
}

export function attachRewardsToSession(session, data) {
  const version = data.version ?? 1;
  if (data.rewards) {
    session.rewards = { ...defaultRewardsState(), ...data.rewards };
    syncRewardsFlags(session);
    return;
  }
  migrateSessionRewards(session, version);
}
